### PR TITLE
Some performance enhancements to model.py

### PR DESCRIPTION
**Pull request by groutr**
Most of the enhancements are the result of removing extra dictionary copies/creations.

1. Improve dictionary handling in `.from_dict()` methods. If passing kwargs to method, the new methods are faster.
2. Remove dictionary unpackings where they are unnecessary.
3. Create a _GEO_TYPES dictionary to avoid the creation of a list in decode()

### CODE REVIEW

**Comment by groutr** [comment]
A comparison of the new/old approaches in `.from_dict()`, where m is a mapping of length 8.
```
In [50]: %timeit from_dict2(**m)  # new
695 ns ± 11 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [51]: %timeit from_dict(**m)  # old
1.15 µs ± 6.93 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

**Review by mwtoews — APPROVED** [review]
LGTM, thanks! I'll merge in a few more days (in case there is other feedback).

File: fiona/model.py [hunk 1]
```diff
@@ -194,15 +198,19 @@ def __init__(self, coordinates=None, type=None, geometries=None, **data):
 
     @classmethod
     def from_dict(cls, ob=None, **kwargs):
-        data = dict(getattr(ob, "__geo_interface__", ob) or {}, **kwargs)
+        if ob is not None:
+            data = dict(getattr(ob, "__geo_interface__", ob))
+            data.update(kwargs)
+        else:
+            data = kwargs
```
[the remark this answers is not in the document]
Comment (sgillies) [reply fiona/model.py:205]: I've benchmarked the two forms using the following.

```python
# dictbm.py
import timeit

import shapely

ob = shapely.Point(0, 0)
kwargs = {"a": 1, "b": "hi"}

s = "dict(getattr(ob, '__geo_interface__', ob) or {}, **kwargs)"

print(timeit.timeit(stmt=s, setup="from __main__ import ob, kwargs", number=100000))

s = """\
if ob is not None:
    data = dict(getattr(ob, '__geo_interface__', ob))
    data.update(kwargs)
else:
    data = kwargs
"""

print(timeit.timeit(stmt=s, setup="from __main__ import ob, kwargs", number=100000))
```

They have the same performance. So, let's keep the single line form.

```
$ python dictbm.py 
0.6029997089644894
0.6092167729511857
```

Maybe the performance gain is all about not creating a new list of geometry types each time?

**Review by sgillies — CHANGES_REQUESTED** [review]

File: fiona/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (groutr) [reply fiona/model.py:205]: Accessing `__geo_interface__` is the slow part. If ob is None, then we don't have to check, which is what my approach attempts to do.

[the remark this answers is not in the document]
Comment (sgillies) [reply fiona/model.py:205]: @groutr I ran Python's bytecode disassembler on the functions and there are differences that I didn't expect. Our old code has extra BUILD_MAP and BUILD_TUPLE ops and a DICT_MERGE, all with some extra cost and overhead. Let's go with your new code!

```
>>> def func(ob, **kwargs):
...     return dict(getattr(ob, '__geo_interface__', ob) or {}, **kwargs)
... 
>>> dis.dis(func)
  1           0 RESUME                   0

  2           2 LOAD_GLOBAL              1 (NULL + dict)
             14 LOAD_GLOBAL              3 (NULL + getattr)
             26 LOAD_FAST                0 (ob)
             28 LOAD_CONST               1 ('__geo_interface__')
             30 LOAD_FAST                0 (ob)
             32 PRECALL                  3
             36 CALL                     3
             46 JUMP_IF_TRUE_OR_POP      1 (to 50)
             48 BUILD_MAP                0
        >>   50 BUILD_TUPLE              1
             52 BUILD_MAP                0
             54 LOAD_FAST                1 (kwargs)
             56 DICT_MERGE               1
             58 CALL_FUNCTION_EX         1
             60 RETURN_VALUE
```

```
>>> def func2(ob, **kwargs):                                                                         
...     if ob is not None:                        
...         data = dict(getattr(ob, '__geo_interface__', ob))
...         data.update(kwargs)           
...     else:                                     
...         data = kwargs                                                                            
...     return data                                                                                  
...                                               
>>> dis.dis(func2)                                              
  1           0 RESUME                   0     
                                                  
  2           2 LOAD_FAST                0 (ob)
              4 POP_JUMP_FORWARD_IF_NONE    52 (to 110)
                                                  
  3           6 LOAD_GLOBAL              1 (NULL + dict)
             18 LOAD_GLOBAL              3 (NULL + getattr)
             30 LOAD_FAST                0 (ob)                                                      
             32 LOAD_CONST               1 ('__geo_interface__')
             34 LOAD_FAST                0 (ob)
             36 PRECALL                  3
             40 CALL                     3
             50 PRECALL                  1
             54 CALL                     1
             64 STORE_FAST               2 (data)

  4          66 LOAD_FAST                2 (data)
             68 LOAD_METHOD              2 (update)
             90 LOAD_FAST                1 (kwargs)
             92 PRECALL                  1
             96 CALL                     1
            106 POP_TOP
            108 JUMP_FORWARD             2 (to 114)

  6     >>  110 LOAD_FAST                1 (kwargs)
            112 STORE_FAST               2 (data)

  7     >>  114 LOAD_FAST                2 (data)
            116 RETURN_VALUE
```

**Review by sgillies — APPROVED** [review]
@groutr thank you!

**Comment by sgillies** [comment]
I'm going to merge this and cherry pick for 1.9.4.

**Pull request merged**